### PR TITLE
Update resource reservation end date

### PR DESCRIPTION
**Pull request by victor-champonnois**
[Task](https://gestion.coopiteasy.be/web#id=7527&view_type=form&model=project.task&action=475&active_id=198)

[Related PR](https://github.com/coopiteasy/cie-custom/pull/47)

- create set_activity_as_done method that updates the state and the reservation end date
- call this method instead of just updating the state

### CODE REVIEW

File: resource_activity/models/resource_activity.py [hunk 1]
```diff
@@ -368,12 +368,17 @@ def action_done(self):
...
+                activity.set_activity_as_done()
             else:
                 raise ValidationError(
                     _("You can only set the activity to done from Sale state.")
                 )
 
+    def set_activity_as_done(self):
+        self.state = "done"
+        if self.date_end > datetime.now():
+            self.registrations.allocations.date_end = datetime.now()
```
Comment (robinkeunen) [inline resource_activity/models/resource_activity.py]: I think this won't work if there is more than one registration on the activity or more than one allocation on the registration.

**Review by robinkeunen — CHANGES_REQUESTED** [review]
Can you add some unit tests as well ?

File: resource_activity/models/resource_activity.py (hunk 1 above)
Comment (robinkeunen) [reply resource_activity/models/resource_activity.py]: Also, i'd add `api.multi` and `self.ensure_one()`

**Review by robinkeunen — CHANGES_REQUESTED** [review]

File: resource_activity/models/resource_activity.py (hunk 1 above)
Comment (victor-champonnois) [reply resource_activity/models/resource_activity.py]: Fixed, I also refactored to simplify the tests

**Comment by victor-champonnois** [comment]
- added unit tests
- refactored to simplify testing (avoid depending on datetime.now after reading [this](https://hakibenita.com/python-dependency-injection))
- adding support for multiple registrations/allocations
- adding `@api.multi` and `ensure_one()`

File: resource_activity/tests/test_resource_activity.py [hunk 2]
```diff
@@ -114,3 +114,93 @@ def test_activity_w_booked_resources(self):
...
+                # set by _onchange_allocation_end in real life
+                "resource_allocation_end": date_end,
+                "location_id": self.main_location.id,
+                "activity_type": self.activity_type.id,
+                "registrations": [(0, 0, registration_1), (0, 0, registration_2)],
+            }
+        )
+        activity_not_finished.search_all_resources()
+        activity_not_finished.reserve_needed_resource()
+        activity_not_finished.update_resource_registration_date_end(date_done)
```
Comment (robinkeunen) [inline resource_activity/tests/test_resource_activity.py]: Shouldn't you rather close the activity and test that `set_activity_as_done` and `update_resource_registration_date_end` were called ?

**Review by robinkeunen — CHANGES_REQUESTED** [review]
Big Thank you for the unit tests. I'd go a step further and test each case in a separate unit test.

Also, very interesting note about date injection.

```python
class TestResourceActivity(test_base.TestResourceActivityBase):
    def setUp(self):
        super().setUp()
        # prepare common setup data 
        date_start = datetime(2022, 2, 1, 1, 1, 1)
        registration_1 = {
            "attendee_id": self.partner_demo.id,
            "quantity": 1,
            "quantity_needed": 1,
            "booking_type": "booked",
            "resource_category": self.mtb_category.id,
            "product_id": self.bike_product.id,
        }
        ...

    def test_activity_ended_early_updates_registrations_date_end(self):
    
        date_start = datetime(2022, 2, 1, 1, 1, 1)
        date_end = date_start + timedelta(days=4)
        date_done = date_start + timedelta(days=2)
        ...
        
        for registration_date_end in activity_not_finished.registrations.mapped(
            "allocations").mapped("date_end"):
            self.assertEqual(registration_date_end, date_done)
        activity_not_finished.unreserve_resources()
        
    
    def test_activity_set_as_done_after_end_does_not_updates_registrations_date_end
        date_end = date_start + timedelta(days=4)
        date_done = date_start + timedelta(days=6)
        ...
        for registration_date_end in activity_already_finished.registrations.mapped("allocations").mapped("date_end"):
            self.assertEqual(registration_date_end, date_end)
```


You'll see that I use lengthy descriptive names for the functions : they only exixt here.

Some day, I would like to implement [these best practices as well](https://www.youtube.com/watch?v=mks7haGqmKY) but I have yet to get familiar w/ them.

**Comment by robinkeunen** [comment]
I see my comment in the code breaks your suggestion to use dependency injection. Not sure what's the best way to go.

File: resource_activity/tests/test_resource_activity.py (hunk 2 above)
Comment (victor-champonnois) [reply resource_activity/tests/test_resource_activity.py]: Yes you are right, see what I did in my answer below

**Comment by victor-champonnois** [comment]
- refactored the tests
- added tests to check that `action_done` changes the state and changes the date_end. 
    - Because of the use of datetime_now() in action_done, I could not precisely test for the date change so I just tested that `datetime.now()` is greater than the updated date (assertAlmostEqual does not seem to work with datetime). The precise check is done with the first test. 
    - Also did not succeed to use `mock` in order to mock `datetime` (odoo modules might not behave as normal python package ?)
    - We could consider using `freezegun` as explained in the video

**Review by robinkeunen — APPROVED** [review]
Super travail => it'd be nice if you could share your findings about date injection at the CC.

**Comment by victor-champonnois** [comment]
Using freezegun for tests

**Pull request merged**